### DISCUSSION TITLE
Is there a way to ignore `AVG_ROW_LENGTH` in `skeema diff`

### DISCUSSION THREAD

**Discussion by bravecorvus** [skeema/skeema#236] [Q&A] [answered]
I use `skeema` to generate diffs, and I notice the `AVG_ROW_LENGTH`.

Is there a way to disable this behavior?

The current command I use is `skeema diff --allow-unsafe > ./diff.sql`

**Comment by evanelias** [comment, marked as answer]
Skeema's diffs are based on what you have in your *.sql files, and what your database contains (obtained via `information_schema`, `SHOW CREATE TABLE`, etc). So if the diff contains `AVG_ROW_LENGTH` clauses, it is because you have those clauses either in your *.sql files, or in your actual database tables.

If you don't want part of the diff, the best solution would be to address it directly in either the db or the *.sql files. But alternatively you could pipe the diff output to another program like `sed` which strips the unwanted text using a regex.

What storage engine are these tables? If I recall correctly, `AVG_ROW_LENGTH` is only used in MyISAM, not InnoDB. Skeema primarily supports InnoDB.

**Comment by bravecorvus** [comment]
We manage a staging and production version of our databases, and our staging one is based on AWS RDS Aurora MySQL Serverless v2 (MySQL 8.0) while our production database is based on traditional AWS RDS Aurora MySQL traditional (MySQL 5.7).

The differences are likely due to diferences in MySQL version. Thanks

**Comment by evanelias** [comment]
This isn't version dependent. Again, AVG_ROW_LENGTH attribute only has an effect on MyISAM tables. Typically you'd only see AVG_ROW_LENGTH attributes on InnoDB tables if they were converted from MyISAM at some point in the past. That said, if you create or alter an InnoDB table explicitly using this clause, MySQL will "remember" the clause in `information_schema` and `SHOW CREATE TABLE` even though the clause has no effect.

Assuming your tables are actually InnoDB, you can permanently remove those clauses from your tables via `ALTER TABLE ... AVG_ROW_LENGTH=0`. This is what Skeema will auto-generate for you, if you remove those clauses from your *.sql files manually and then diff against a DB that has those clauses.

**Comment by bravecorvus** [comment]
Thanks for the clarification. I should specify that AWS Aurora MySQL only promises "MySQL compatible", but not "full fledged MySQL server". And I am unable to find any definative documentation on the default behavior prior to 2021, but I assume InnoDB is default.

I can definitely say we did not convert MyISAM tables into InnoDB tables.

But in regards to the RDS Aurora Serverless v2 cluster, it is definitely a post-2021 MySQL 8.0 compatible Aurora instance, meaning it is likely affected by the [change in default internal temporary table storage engine to TempTable](https://aws.amazon.com/blogs/database/use-the-temptable-storage-engine-on-amazon-rds-for-mysql-and-amazon-aurora-mysql/).

Do you think the behavior of the proprietary internal temp table storage engine could be the cause of this?

**Comment by evanelias** [comment]
No, it's not related to temp tables or the internal temp table storage engine. Skeema does not interact with temp tables.

For real (non-temp) tables, Aurora only supports InnoDB.

As I said previously, if someone creates an InnoDB table with a AVG_ROW_LENGTH clause, MySQL will "remember" it and expose it in `information_schema` and `SHOW CREATE TABLE`, even though the clause has no effect whatsoever. This is most likely what happened here.

Again, you can remove these clauses from your tables via `ALTER TABLE ... AVG_ROW_LENGTH=0`. This is what Skeema will auto-generate for you, if you remove those clauses from your *.sql files manually and then diff against a DB that has those clauses.
